hnn_core/parallel_backends.py

```diff
@@ -0,0 +1,265 @@
+        hyperthreading = False
+        if self.n_cores > physical_cores:
+            hyperthreading = True
+
+        self.mpi_cmd_str = mpi_cmd
+
+        if self.n_cores > 1:
+            print("MPI will run over %d cores" % (self.n_cores))
+        else:
+            print("Only have 1 core available. Running simulation without MPI")
+            print("Consider using Joblib_backend with n_jobs > 1 "
+                  "for running multiple trials")
+            return
+
+        if self.n_jobs > 1:
+            raise ValueError("Nested parallelism is not currently supported"
+                             " with MPI_backend!\n"
+                             "Please use joblib for embarassinly parallel jobs"
+                             " (n_jobs > 1)\n"
+                             "or multiple cores per simulation with"
+                             " MPI_backend\n")
+
+        if hyperthreading:
+            self.mpi_cmd_str += ' --use-hwthread-cpus'
+
+        if oversubscribe:
+            self.mpi_cmd_str += ' --oversubscribe'
+
+        self.mpi_cmd_str += ' -np ' + str(self.n_cores)
+
+        self.mpi_cmd_str += ' nrniv -python -mpi -nobanner ' + \
+            os.path.join(os.path.dirname(hnn_core.__file__), 'mpi_child.py')
+
+    def __enter__(self):
+        global _backend
+        self._old_backend = _backend
+        _backend = self
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        global _backend
+        _backend = self._old_backend
+
+    def simulate(self, net):
+        """Simulate the HNN model in parallel on all cores
+
+        Parameters
+        ----------
+        net : Network object
+            The Network object specifying how cells are
+            connected.
+
+        Returns
+        -------
+        dpl: list of Dipole
+            The Dipole results from each simulation trial
+        """
+        from subprocess import Popen, PIPE
+        import shlex
+        import pickle
+        import codecs
+        import os
+        from sys import platform
+
+        # just use the joblib backend for a single core
+        if self.n_cores == 1:
+            return Joblib_backend(n_jobs=1).simulate(net)
+
+        print("Running %d trials..." % (net.params['N_trials']))
+
+        # Split the command into shell arguments for passing to Popen
+        cmdargs = shlex.split(self.mpi_cmd_str, posix="win" not in platform)
+
+        pickled_params = codecs.encode(pickle.dumps(net.params),
+                                       "base64").decode()
+
+        # set some MPI environment variables
+        my_env = os.environ.copy()
+        my_env["OMPI_MCA_btl_base_warn_component_unused"] = '0'
+        # Start the simulation in parallel!
+        proc = Popen(cmdargs, stdin=PIPE, stdout=PIPE, stderr=PIPE, env=my_env,
+                     cwd=os.getcwd(), universal_newlines=True)
+
+        # wait until process completes
+        out, err = proc.communicate(pickled_params)
```

Comment:
I tried to get real-time output from the MPI child process while it is running, but ran into the same deadlock issues mention in the `subprocess` documentation as reasons why they recommend `communicate`. There might be a workaround, but it'll be complicated, and good for another PR.

Reply:
sounds good!